modules/core/include/opencv2/core/dualquaternion.inl.hpp

```diff
@@ -0,0 +1,392 @@
+    return createFromQuat(r, trans * r / 2);
+}
+
+template <typename T>
+DualQuat<T> DualQuat<T>::createFromMat(InputArray _R)
+{
+    CV_CheckTypeEQ(_R.type(), cv::traits::Type<T>::value, "");
+
+    Mat R = _R.getMat();
+    Quat<T> r = Quat<T>::createFromRotMat(R.colRange(0, 3).rowRange(0, 3));
+    Quat<T> trans(0, R.at<T>(0, 3), R.at<T>(1, 3), R.at<T>(2, 3));
+    return createFromQuat(r, trans * r / 2);
+}
+
+template <typename T>
+DualQuat<T> DualQuat<T>::createFromPitch(const T angle, const T d, const Quat<T> &axis, const Quat<T> &moment)
+{
+    T half_angle = angle / 2, half_d = d / 2;
+    Quat<T> qaxis = axis.normalize();
+    Quat<T> dual = Quat<T>(-half_d * std::sin(half_angle), 0.0, 0.0, 0.0) + std::sin(half_angle) * moment +
+        half_d * std::cos(half_angle) * qaxis;
+    return createFromQuat(Quat<T>::createFromAngleAxis(angle, Vec<T, 3>{qaxis[1], qaxis[2], qaxis[3]}), dual);
+}
+
+template <typename T>
+inline bool DualQuat<T>::operator==(const DualQuat<T> &q) const
+{
+    return (abs(w - q.w) < CV_DUAL_QUAT_EPS && abs(x - q.x) < CV_DUAL_QUAT_EPS &&
+            abs(y - q.y) < CV_DUAL_QUAT_EPS && abs(z - q.z) < CV_DUAL_QUAT_EPS &&
+            abs(w_ - q.w_) < CV_DUAL_QUAT_EPS && abs(x_ - q.x_) < CV_DUAL_QUAT_EPS &&
+            abs(y_ - q.y_) < CV_DUAL_QUAT_EPS && abs(z_ - q.z_) < CV_DUAL_QUAT_EPS);
+}
+
+template <typename T>
+inline Quat<T> DualQuat<T>::getRealQuat() const
+{
+    return Quat<T>(w, x, y, z);
+}
+
+template <typename T>
+inline Quat<T> DualQuat<T>::getDualQuat() const
+{
+    return Quat<T>(w_, x_, y_, z_);
+}
+
+template<typename T>
+inline DualQuat<T> DualQuat<T>::conjugate() const
+{
+    return DualQuat<T>(w, -x, -y, -z, w_, -x_, -y_, -z_);
+}
+
+template <typename T>
+DualQuat<T> DualQuat<T>::norm(QuatAssumeType assumeUnit) const
+{
+    Quat<T> real = getRealQuat();
+    T realNorm = real.norm();
+    if (assumeUnit)
+    {
+        return DualQuat<T>(realNorm, 0, 0, 0, 0, 0, 0, 0);
+    }
+    Quat<T> dual = getDualQuat();
+    if (realNorm < CV_DUAL_QUAT_EPS){
+        return DualQuat<T>(0, 0, 0, 0, 0, 0, 0, 0);
+    }
+    return DualQuat<T>(realNorm, 0, 0, 0, real.dot(dual) / realNorm, 0, 0, 0);
+}
+
+template <typename T>
+inline Quat<T> DualQuat<T>::getRotation(QuatAssumeType assumeUnit) const
+{
+    if (assumeUnit)
+    {
+        return getRealQuat();
+    }
+    return getRealQuat().normalize();
+}
+
+template <typename T>
+Quat<T> DualQuat<T>::getTranslation(QuatAssumeType assumeUnit) const
+{
+    if (assumeUnit)
+    {
+        return 2.0 * getDualQuat() * getRealQuat().conjugate();
+    }
+    DualQuat<T> q = normalize();
+    return 2.0 * q.getDualQuat() * q.getRealQuat().conjugate();
+}
+
+template <typename T>
+inline DualQuat<T> DualQuat<T>::normalize() const
+{
+    DualQuat<T> qNorm = norm();
+    if (qNorm.w < CV_DUAL_QUAT_EPS)
+    {
+        CV_Error(Error::StsBadArg, "Cannot normalize this dual quaternion: the norm is too small.");
+    }
+    return *this / qNorm;
```

Comment:
While this code is elegant and simple, another version can be used to avoid a chain of calls `norm()`, `operator/()`, `inv()`, `mul()`:
```
a_nr = a/norm(a),  b_nr = b/norm(a)
normalized = a_nr + e*(b_nr - a_nr*dot(a_nr, b_nr))
```

Reply:
OK, thanks.